executor/common_linux.h

```diff
@@ -1626,6 +1626,377 @@ static long syz_genetlink_get_family_id(volatile long name)
+
+int hci_open_dev(int dev_id)
+{
+	struct sockaddr_hci a;
+	int dd, err;
+
+	/* Create HCI socket */
+	dd = syz_init_net_socket(AF_BLUETOOTH, SOCK_RAW | SOCK_CLOEXEC, BTPROTO_HCI);
+	if (dd < 0)
+		return dd;
+
+	/* Bind socket to the HCI device */
+	memset(&a, 0, sizeof(a));
+	a.hci_family = AF_BLUETOOTH;
+	a.hci_dev = dev_id;
+	if (bind(dd, (struct sockaddr*)&a, sizeof(a)) < 0)
+		goto failed;
+
+	return dd;
+
+failed:
+	err = errno;
+	close(dd);
+	errno = err;
+
+	return -1;
+}
+
+static int vhci_fd = -1;
+
+int hci_send_event_packet(int fd, uint8 evt, void* data, size_t data_len)
+{
+	struct iovec iv[3];
+
+	hci_event_hdr hdr;
+	hdr.evt = evt;
+	hdr.plen = data_len;
+
+	uint8 type = HCI_EVENT_PKT;
+
+	iv[0].iov_base = &type;
+	iv[0].iov_len = sizeof(type);
+	iv[1].iov_base = &hdr;
+	iv[1].iov_len = HCI_EVENT_HDR_SIZE;
+	iv[2].iov_base = data;
+	iv[2].iov_len = data_len;
+
+	return writev(fd, iv, sizeof(iv) / sizeof(struct iovec));
+}
+
+int hci_send_event_cmd_complete(int fd, uint16 opcode, void* data, size_t data_len)
+{
+	struct iovec iv[4];
+
+	hci_event_hdr hdr;
+	hdr.evt = EVT_CMD_COMPLETE;
+	hdr.plen = EVT_CMD_COMPLETE_SIZE + data_len;
+
+	evt_cmd_complete evt_hdr;
+	evt_hdr.ncmd = 1;
+	evt_hdr.opcode = opcode;
+
+	uint8 type = HCI_EVENT_PKT;
+
+	iv[0].iov_base = &type;
+	iv[0].iov_len = sizeof(type);
+	iv[1].iov_base = &hdr;
+	iv[1].iov_len = HCI_EVENT_HDR_SIZE;
+	iv[2].iov_base = &evt_hdr;
+	iv[2].iov_len = EVT_CMD_COMPLETE_SIZE;
+	iv[3].iov_base = data;
+	iv[3].iov_len = data_len;
+
+	return writev(fd, iv, sizeof(iv) / sizeof(struct iovec));
+}
+
+static int vhci_init = 0;
+
+// TODO: fuzz event packets.
+void process_command_pkt(int fd, char* buf, size_t buf_size)
+{
+	hci_command_hdr* hdr = (hci_command_hdr*)(buf + 1);
+	uint16 opcode = hdr->opcode;
+	uint16 ogf = cmd_opcode_ogf(opcode);
+	uint16 ocf = cmd_opcode_ocf(opcode);
+
+	if (ogf == OGF_HOST_CTL) {
+		if (ocf == OCF_WRITE_CONN_ACCEPT_TIMEOUT)
+			vhci_init = 1;
+	}
+	if (ogf == OGF_INFO_PARAM) {
+		switch (ocf) {
+		case OCF_READ_BD_ADDR: {
+			read_bd_addr_rp rp;
+			rp.status = 0;
+			memcpy(&rp.bdaddr, "\x66\x55\x44\x33\x22\x11", 6);
+			hci_send_event_cmd_complete(fd, opcode, &rp, READ_BD_ADDR_RP_SIZE);
+			return;
+		}
+		case OCF_READ_BUFFER_SIZE: {
+			read_buffer_size_rp rp;
+			rp.status = 0;
+			rp.acl_mtu = 1021;
+			rp.sco_mtu = 96;
+			rp.acl_max_pkt = 4;
+			rp.sco_max_pkt = 6;
+			hci_send_event_cmd_complete(fd, opcode, &rp, READ_BUFFER_SIZE_RP_SIZE);
+			return;
+		}
+		}
+	}
+
+	char status[0xf9];
+	memset(status, 0, sizeof(status));
+	hci_send_event_cmd_complete(fd, opcode, status, sizeof(status));
+}
+
+#define MAX_EVENTS 1
+#define BUF_SIZE 4096
+
+void* event_thread(void* argp)
+{
+	struct epoll_event events[MAX_EVENTS];
+	int event_count;
+	char buf[BUF_SIZE];
+	size_t buf_size;
+
+	int epoll_fd = *(int*)argp;
+
+	while (1) {
+		event_count = epoll_wait(epoll_fd, events, MAX_EVENTS, -1);
```

Comment:
We add just 1 fd to the wait set. Couldn't we just call read without the epoll? It should be equivalent, no?

Reply:
True, I copied it from my PoC and did not clean it up yet.